src/compiler/checker.ts

```diff
@@ -3513,72 +3522,79 @@ module ts {
+                let typeArguments = map(node.typeArguments, getTypeFromTypeNode);
+                let id = getTypeListId(typeArguments);
+                let links = getSymbolLinks(symbol);
+                if (!links.instantiations) {
+                    links.instantiations = {};
+                    // The declared type corresponds to an instantiation with its own type parameters as type arguments
+                    links.instantiations[getTypeListId(typeParameters)] = type;
                 }
+                return links.instantiations[id] || (links.instantiations[id] = instantiateType(type, createTypeMapper(typeParameters, typeArguments)));
             }
-            else {
-                if (typeArguments) {
-                    error(node, Diagnostics.Type_0_is_not_generic, typeToString(type));
-                    return undefined;
-                }
+            if (node.typeArguments) {
+                error(node, Diagnostics.Type_0_is_not_generic, symbolToString(symbol));
+                return unknownType;
             }
-
             return type;
         }
 
+        function getTypeFromTypeParameterReference(node: TypeReferenceNode | ExpressionWithTypeArguments, symbol: Symbol): Type {
+            // TypeScript 1.0 spec (April 2014): 3.4.1
+            // Type parameters declared in a particular type parameter list
+            // may not be referenced in constraints in that type parameter list
+            // Implementation: such type references are resolved to 'unknown' type that usually denotes error
+            return isTypeParameterReferenceIllegalInConstraint(node, symbol) ? unknownType : getDeclaredTypeOfSymbol(symbol);
+        }
+
+        function getTypeFromTypeReference(node: TypeReferenceNode | ExpressionWithTypeArguments): Type {
+            let links = getNodeLinks(node);
+            if (!links.resolvedType) {
+                // We only support expressions that are simple qualified names. For other expressions this produces undefined.
                let typeNameOrExpression = node.kind === SyntaxKind.TypeReference ? (<TypeReferenceNode>node).typeName :
+                    isSupportedExpressionWithTypeArguments(<ExpressionWithTypeArguments>node) ? (<ExpressionWithTypeArguments>node).expression :
+                    undefined;
+                let symbol = typeNameOrExpression && resolveEntityName(typeNameOrExpression, SymbolFlags.Type) || unknownSymbol;
+                let type = symbol.flags & (SymbolFlags.Class | SymbolFlags.Interface) ? getTypeFromClassOrInterfaceReference(node, symbol) :
+                    symbol.flags & SymbolFlags.TypeParameter ? getTypeFromTypeParameterReference(node, symbol) :
+                    symbol.flags & SymbolFlags.TypeAlias ? getTypeFromTypeAliasReference(node, symbol) :
+                    getDeclaredTypeOfSymbol(symbol);
+                // Cache both the resolved symbol and the resolved type. The resolved symbol is needed in when we check the

```

Comment:
I'm not surprised. It is not an obvious case to test. Can you add tests for this as well?